<a href="https://colab.research.google.com/github/ShreyJais/Speech-Processing/blob/main/2348558_SPR_lab8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Viterbi Dynamic Programming Algorithm to Find the Most Likely Sequence of Hidden States



In [ ]:
import numpy as np
import matplotlib.pyplot as plt

The Viterbi algorithm is a dynamic programming algorithm used to find the most probable sequence of hidden states given an observation sequence. In this case, the hidden states represent the phonemes (/h/, /e/, /l/, /o/) and the observations represent the acoustic feature vectors.

In [ ]:
states = ['h', 'e', 'l', 'o']
observations = ['O1', 'O2', 'O3', 'O4']

A = np.array([ #Transition Probability Matrix (A)
    [0.0, 0.7, 0.3, 0.0],
    [0.0, 0.2, 0.6, 0.2],
    [0.0, 0.0, 0.3, 0.7],
    [0.0, 0.0, 0.1, 0.9]
])

B = np.array([ #Emission Probability Matrix (B)
    [0.6, 0.2, 0.1, 0.1],
    [0.1, 0.7, 0.1, 0.1],
    [0.1, 0.1, 0.6, 0.2],
    [0.2, 0.1, 0.2, 0.5]
])

pi = np.array([1.0, 0.0, 0.0, 0.0]) # nitial probabilities (π)

observation_sequence = ['O1', 'O2', 'O3', 'O4']
observation_index = {'O1': 0, 'O2': 1, 'O3': 2, 'O4': 3}

- States: We have 4 phonemes: /h/, /e/, /l/, /o/.
- Observations: The observations are feature vectors: O1, O2, O3, O4.
- Transition Matrix (A): Represents the probabilities of moving from one state to another.
- Emission Matrix (B): Represents the probabilities of emitting an observation given a state.
- Initial Probabilities (π): Represents the probability of starting at each phoneme (only /h/ starts with probability 1).

In [ ]:
def viterbi_algorithm():
    n_states = len(states)
    n_observations = len(observations)
    viterbi_matrix = np.zeros((n_observations, n_states))
    backpointer_matrix = np.zeros((n_observations, n_states), dtype=int)
    for s in range(n_states):
        viterbi_matrix[0][s] = pi[s] * B[s][observation_index[observation_sequence[0]]]
    for t in range(1, n_observations):
        for s in range(n_states):
            max_prob = -1
            max_state = -1
            for prev_s in range(n_states):
                prob = viterbi_matrix[t-1][prev_s] * A[prev_s][s] * B[s][observation_index[observation_sequence[t]]]
                if prob > max_prob:
                    max_prob = prob
                    max_state = prev_s
            viterbi_matrix[t][s] = max_prob
            backpointer_matrix[t][s] = max_state
    best_path = np.zeros(n_observations, dtype=int)
    best_path[-1] = np.argmax(viterbi_matrix[-1])
    for t in range(n_observations - 2, -1, -1):
        best_path[t] = backpointer_matrix[t + 1][best_path[t + 1]]
    best_path_phonemes = [states[state] for state in best_path]
    best_sequence_probability = viterbi_matrix[-1][best_path[-1]]

    return best_path_phonemes, best_sequence_probability

- **Initialization**: We initialize the viterbi_matrix for the first observation, using the initial probabilities and emission probabilities.
- **Recursion**: For each subsequent observation, we calculate the most probable state path by looking at the previous state, the transition probability, and the emission probability.
- **Backtracking**: After filling out the viterbi_matrix, we backtrack to find the most probable sequence of states.
- **Probability**: The probability of the most likely sequence is the value at the last observation for the most likely state.

In [ ]:
best_sequence, best_probability = viterbi_algorithm()

print("Most Likely Phoneme Sequence:", best_sequence)
print("Probability of the Most Likely Sequence:", best_probability)

Most Likely Phoneme Sequence: ['h', 'e', 'l', 'o']
Probability of the Most Likely Sequence: 0.03704399999999999


### Inference


The Viterbi algorithm confirms that the sequence /h/ → /e/ → /l/ → /o/ is the most probable hidden phoneme sequence corresponding to the observed acoustic feature vectors [O1, O2, O3, O4].

The computed probability of this sequence (0.03704) reflects the joint likelihood of:

- Starting at /h/,
- Transitioning between states based on the transition probabilities,
- Emitting the corresponding observations based on emission probabilities.